### ISSUE TITLE
Support offsets on windowing

### ISSUE THREAD

**Issue by big-andy-coates** [confluentinc/ksql#4460]
KSQL aligns windows to UTC.  This can mean that `1 DAY` windows don't aline with actual days when processing data in other timezones.

KSQL should support the ability to set either an offset or a timezone to allow users to customize where the boundaries lie.   This will be much easier if we first switch [windowing to table function primitives](#4442).

e.g.

```
SELECT * FROM TUMBLING(
       source => FOO,
       size => '1 MINUTE'
       offset => '42 SECONDS'
) ...
```

Would have window start times of 12:00:42, 12:01:42, etc.

Or

```
SELECT * FROM TUMBLING(
       source => FOO,
       size => '1 DAY'
       timezone => 'PSD'
)...
```

Would have window start times alined to midnight in the specified timezone.

**Comment by blueedgenick** [comment]
not sure that really needs such a drastic change :)    Mechanically of
course it boils down to some timestamp-extractor-juggling. Perhaps it could
be specified in terms of timezone rather than some "offset", seems that
would be easier to reason about and specify ?
something like `select ......window tumbling(size 1 day, timezone
'GMT').....` could be a minimally-invasive and easily-grokkable way to do
this.  I specifically use a "1 day" example as that's where i've seen this
come  up most often - folks want daily windows aligned to their local
midnight.

**Comment by k1th** [comment]
we're going to create a PR for a UDF that adjusts an epoch to the same epoch offset to a specific time zone. The resulting epoch can then be used to align windows to days in that TZ.

**Comment by big-andy-coates** [comment]
@blueedgenick of course  we can just add more syntax to the current windowing.  Think I was getting over excited when I raised this issue ;)

@k1th nice. It's worth noting that epoch isn't TZ sensitive, as it's _always_ UTC.  So you're UDF can't really be converting epoch's between timezones. You're UDF would actually be returning a _different_ point in time.  This would make it perfect to correct bad data, i.e. data where a date string was converted into an epoch using the wrong tz. But would mean using it to correct windowing boundaries may lead to unexpected side effects downstream. Definitely still useful, and definitely a potential workaround for this issue, but not a perfect solution.

**Comment by k1th** [comment]
@big-andy-coates 
yeah, we know that epoch is always UTC. In our case , the resulting records are then „in the future“ due go being east of 0. There is no other solution we could come up with unless the window clause had another param specifying an offset.
Also, grace periods are affected (by max. 12 hours) due to the different time as seen by Streams.

**Comment by big-andy-coates** [comment]
Cool. Look forward to the PR!

**Comment by qingsonggithub** [comment]
hi @k1th @vcrfxia is this issue any progress?  as i know , flink is support set offset in windows， detail in this [link](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/operators/windows/#tumbling-windows)
And in my case, my window size is 1 hour and  i want to make the window start from 10:30 , say first window is [10:30 ,11:30] not  [10:00 ,11:00] , if window don't support offset, i can't get window start from 10:30 .

**Comment by k1th** [comment]
@qingsonggithub As far as I know, there are no news about the window boundary. 
We still use the time offset functionality now included with KSQL to change the time before windowing and then interpreting the windows correctly afterwards.

**Comment by mjsax** [comment]
We will first need a KIP for Kafka Streams to add this functionality there. Afterwards, ksqlDB can use it.
